cpp/docs/TESTING.md

```diff
@@ -0,0 +1,408 @@
+# Unit Testing in libcudf
```

Comment:
I think it would be worth having a best practices section for testing.  There are a lot of classic things that are worthy of testing which could be easily overlooked:

- Anything that involves manipulating bitmasks (especially hand-rolled kernels) should have tests that check : # rows < 32, # rows == 32, # rows > 32 and # rows > 64.   Lots of ways to flub those in kernels.

- For hand-rolled kernels, large enough inputs that they span multiple grid blocks.

- For anything involving strings or lists,  exhaustive combinations of empty strings/lists, null strings/lists and strings/lists with null elements.  

- Tests that explicitly use sliced columns.  That is, columns that have an offset > 0.   An easy to forget case.

- Tests that verify against various forms of "degenerate" column inputs.  Some typical cases are:  empty string columns that have no children (not many paths in cudf can generate these but it does happen);  columns that are of size 0 but somehow have non-null data pointers;  struct columns with no children.

- Decimal types are _not_ included with NumericTypes (but they are included with FixedWidthTypes). Be careful here.

Probably a decent number of other examples.

Reply:
Done. Thanks!